### DISCUSSION TITLE
[New] Quelque chose autour des SPA

### DISCUSSION THREAD

**Discussion by devcrafting** [cnumr/best-practices#208] [Bonnes pratiques] [answered]
Je ne vois nulle part React ou Angular cités alors qu'il semble y avoir pas mal de BP autour de la manipulation de DOM...

[DISCLAIMER] Je ne suis pas un dev front, mais il me semble qu'il faudrait proposer quelque chose autour de ce sujet des frameworks manipulant le DOM qui sont quasi inévitables aujourd'hui...en lien avec les BP de manipulation du DOM (que j'ai du mal à juger) : https://github.com/cnumr/best-practices/blob/main/chapters/BP_041_fr.md,  https://github.com/cnumr/best-practices/blob/main/chapters/BP_042_fr.md, https://github.com/cnumr/best-practices/blob/main/chapters/BP_054_fr.md

**Comment by ACTLEM** [comment]
@EmmanuelDemey un avis de spécialiste ^^? 
Côté React, je n'ai pas encore d'avis fort sur l'intérêt du DOM Virtuel car je n'ai toujours pas creusé si c'est vraiment beaucoup mieux que la manipulation direct du DOM (même si en théorie cela est pus efficace). Ensuite, il est de plus en plus facile de faire du SSR (ou hybride) ce qui réduit cette manipulation côté client.
A voir si parler de DOM Virtuel peut être intéressant dans l'une des BP concernées. Après, je reste toujours embêté par la taille des assets de React (même si cela s'améliore).

**Reply by EmmanuelDemey** [reply]
Je pense qu'il y a un juste milieu entre la taille de l'appli à développer et le choix ou pas d'un framework. 

Je ne choisirai pas un framework pour tous les projets. Mais si c'est un projet "conséquent", je ne cacherai pas dessus. 

Ensuite quel framework... L'éternel débat :) ... Qui (je pense) on ne devrait pas avoir. Ils ont tous des qualités et des défauts. 

Certaines bonnes pratiques sont tout de mêmes communes à n'importe quel fmk 
- Immutabilité
- Rendre le code Tree Shakable
- S'assurer que les composants ne se régénère pas trop souvent 
- Faire du chargement dynamique d'assets
- éviter d'appeler des fonctions qui retourne des objets différents depuis le template


Ai je répondu à la question @ACTLEM  ? :P

**Reply by ACTLEM** [reply]
Merci pour ces éléments qui rejoignent la BP sur une architecture adaptée. A vrai dire la question tourne plus sur:

- Y a-t-il un problème particulier avec les framework et la manipulation du DOM (pour moi pas forcément côté React étant donné le DOM Virtuel)?
- Vois-tu des bonnes pratiques d'écoconception, des intérêts sur les framework ou cela dépend trop du contexte, de l'unité fonctionnelle?

**Reply by arnolem** [reply]
Petite info, il y a des lib comme Svelte qui se vente de ne pas avoir de DOM Virtuel : https://svelte.dev/

Après, pour avoir fait un cas d'usage entre JQUERY et VUEJS sur le projet PingBreak, j'ai clairement vu que modifier 1000 éléments du DOM par seconde ne posait pas de problème à VueJS alors que ce n'était pas la même histoire pour Jquery qui appliquait une à une les 1000 modifications.

Au dela des cas ou React, Angular, ExtJS, ... sont bien utilisés, il y a surtout une armée de jeunes développeurs qui se lance sur le marché en étant des développeurs React ou développeurs VueJS et qui ne jurent que par ça même pour un HelloWorld.
Je pense que le livre devrait tenter de sensibiliser à cela. 
En tout cas, à la lecture de la 3eme édition, j'ai trouvé que l'absence de Framework JS comme React ou d'outils comme WebPack, Gulp, ... rendait ce livre déconnecté de la réalité.

**Reply by devcrafting** [reply]
Cela rejoint ma réponse au commentaire de @fzaninotto : citer ces technologies front inévitables aujourd'hui (à toutes les sauces, sans discernement ?) avec la modération proposée par François ?

**Reply by florinesueur** [reply]
Je n'avais pas vu :) Je pense aussi qu'il faut rajouter quelques techno JS #199

**Reply by EmmanuelDemey** [reply]
Vous avez une liste de BP avec ce que j'ai listé ci dessus.

**Reply by ACTLEM** [reply]
@EmmanuelDemey par rapport aux bonnes pratiques communes des fmk que tu mentionnes, en vois-tu une qui peut être une recommandation sans équivoque?
Il y a peut-être les deux suivantes mais j'ai du mal à trouver un argumentaire fort:

- Rendre le code Tree Shakable
- S'assurer que les composants ne se régénère pas trop souvent

**Reply by EmmanuelDemey** [reply]
Les 5 :)

Immutabilite
- si nous avons des structures de donnees immutables, les frameworks Vont pouvoir facilement vérifier si un paramètre d'un composant a été modifié, en vérifiant juste son adresse, et évitant de faire de la comparaison de chaque propriété. Donc plus rapide, et le composants (et ses sous composants) se gênereront moins souvent. 

Rendre le code Tree Shakable
- Pouvoir supprimer le code de nos dépendances qui ne sont jamais utilisés. Et ainsi réduire la taille du bundle. 


S'assurer que les composants ne se régénère pas trop souvent
- Un peu lié au point sur l'immutabilité. On pourrait mettre aussi la bonne pratique de passer des paramètres Complexe, mais plutôt des primitifs. On pourrait créer une seule.BP sur ce point. 

Faire du chargement dynamique d'assets
- pour réduire la taille du bundle téléchargé. Et ainsi ne proposer que le strict minimum, et de ne pas demander à l'utilisateur de télécharger du JS de pages sur lesquels il n'ira jamais. 

éviter d'appeler des fonctions qui retourne des objets différents depuis le template
- pour voir si la valeur a utiliser est différente, le framework.doit exécuter la fonction a chaque fois qu'il doit régénérer la vue. Si la fonction est coûteuse, et du fait que JS est monothreadé, ça peut bloquer.

**Reply by ACTLEM** [reply]
Merci. J'ai essayé de voir quelle(s) BP(s) pourraient être créée. J'ai un peu de mal surtout pour avoir une BP opérationnelle, pour donner des éléments qui permettent d'affirmer un gain sans pousser à l'utilisation de frameworks (ce qui pourrait être contre-productif dans certains cas).

**Comment by fzaninotto** [comment]
Nous avons effectué quelques mesures avec GreenFrame pour essayer de répondre à cette question. L'idée est de tester la même application en SSR et en SPA. Ce ne sont que des résultats préliminaires sur des petites applications, mais il semble que la différence n'est pas perceptible sur nos tests. 

Voir des détails sur nottre approche dans cet article https://marmelab.com/blog/2021/03/04/argos-comparing-the-energy-consumption-of-two-web-stacks.html#web-server-the-web-stack-does-not-have-a-big-impact-on-the-cpu-usage

Ma compréhension est que la variation de la consommation de ressources côté client dépend grandement du scénario utilisataur. Pour des sites média, où beaucoup d'utilisateurs visitent une seule page depuis Google, charger une SPA est une perte nette. Pour des applications web, où les utilisateurs se logguent et effectuent plusieurs actions, recharger une nouvelle page à chaque action est une perte nette.

Donc il me semble difficile de dire "SPA c'est bien" ou "SPA c'est pas bien" dans l'absolu.

**Reply by ACTLEM** [reply]
Merci @fzaninotto . Cela renforce que c'est trop dépendant de l'UF pour établir une bonne pratique d'écoconception spécifiquement sur les SPA.

**Reply by devcrafting** [reply]
ok peut-être juste un commentaire sur la BP des choix technologiques sur les technologies front pour mettre le résumé de @fzaninotto ;) ?

**Reply by Fred-artwai** [reply]
Il y a quand même une question de fond. Une SPA fait plus travailler le terminal client que le serveur ou du moins qu'un site classique voir Low-tech. Or si on retiens que  les impacts environnement sont majoritairement du côté des terminaux. Et plus on laissera ce genre d'architecture se répandre plus on risque d'encourager les gens à changer de terminal régulièrement.

**Reply by tbroyer** [reply]
Est-ce que ça ne rejoint pas la BP#15? Il y a de vrais cas d'usage pour les SPAs, qui dépendront notamment de l'audience; même si on est bien d'accord que ce n'est pas adapté pour un "site web".

Même si une SPA fait "plus travailler le terminal client", il faut quantifier à partir de quand ça devient problématique. Il faut aussi bien définir ce qu'on appelle SPA: une "petite SPA" pour une UF spécifique sur un site plus large est-elle vraiment problématique ? (par exemple, sur un site de e-commerce, si le passage de commande bascule sur une SPA, est-ce un problème ?)

C'est aussi un compromis avec les capacités du réseau; il y a il me semble des success stories de sites de e-commerce dans les pays "émergents" où une SPA permettait d'obtenir un site utilisable, parce que le réseau (mobile) est tellement mauvais qu'il vaut mieux en demander un peu plus au processeur du terminal pour le rendu et minimiser les échanges réseau (avec une bonne mise en cache et utilisation d'une PWA). D'un point de vue de l'écoconception, minimiser les échanges réseau revient aussi à économiser de la batterie, et donc finalement s'inscrit bien aussi dans l'allongement de la durée de vie du matériel.

**Comment by Fred-artwai** [comment]
Globalement on pourrait mesurer le temps d'exécution de ce genre de technologie avec des mesures JS et proposer une limite a respecter.  Comme avec [cette proposition](https://github.com/cnumr/best-practices/discussions/243).

**Comment by tbroyer** [comment, marked as answer]
Pour info: j'ai évoqué les SPAs dans la nouvelle BP [Ne charger des données/du code que lorsqu'elles sont/il est nécessaire](https://github.com/cnumr/best-practices/blob/main/chapters/BP_4017_fr.md)

**Comment by ACTLEM** [comment]
Le sujet est intéressant mais il n'y a pas de recommandations qui en ressort ou qui apporte une plus value par rapport aux existantes. Je résout donc la discussion qui pourra être réouverte pour des versions ultérieures.
